eChronos spams logs with warn messages if JMX connection towards local C* node is down

Whenever a schedule is trying to run and the JMX connection towards local C* node is down the log is spammed with warn logs.
The method `isUp` in `c.e.b.c.e.core.HostStatesImpl` is called for each vnode which in turn calls `refreshNodeStatus` in `c.e.b.c.e.core.HostStatesImpl`, if the refresh should be performed (there's an interval for this) it will call `tryRefreshHostStates` in `c.e.b.c.e.core.HostStatesImpl` which has a try-catch with a log.warn. This spams logs with the exception below.

```
08:38:32.054 [pool-4-thread-1] WARN  c.e.b.c.e.core.HostStatesImpl - Unable to retrieve host states
java.io.IOException: Failed to retrieve RMIServer stub: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host: localhost; nested exception is: 
        java.net.ConnectException: Connection refused (Connection refused)]
        at javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:369)
        at javax.management.remote.JMXConnectorFactory.connect(JMXConnectorFactory.java:270)
        at com.ericsson.bss.cassandra.ecchronos.connection.impl.LocalJmxConnectionProvider.reconnect(LocalJmxConnectionProvider.java:127)
        at com.ericsson.bss.cassandra.ecchronos.connection.impl.LocalJmxConnectionProvider.getJmxConnector(LocalJmxConnectionProvider.java:75)
        at com.ericsson.bss.cassandra.ecchronos.application.DefaultJmxConnectionProvider.getJmxConnector(DefaultJmxConnectionProvider.java:61)
        at com.ericsson.bss.cassandra.ecchronos.core.JmxProxyFactoryImpl.connect(JmxProxyFactoryImpl.java:66)
        at com.ericsson.bss.cassandra.ecchronos.core.HostStatesImpl.tryRefreshHostStates(HostStatesImpl.java:105)
        at com.ericsson.bss.cassandra.ecchronos.core.HostStatesImpl.refreshNodeStatus(HostStatesImpl.java:87)
        at com.ericsson.bss.cassandra.ecchronos.core.HostStatesImpl.isUp(HostStatesImpl.java:55)
        at com.ericsson.bss.cassandra.ecchronos.core.HostStatesImpl.isUp(HostStatesImpl.java:70)
        at com.ericsson.bss.cassandra.ecchronos.core.repair.state.RepairStateImpl.replicasAreRepairable(RepairStateImpl.java:227)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
        at com.ericsson.bss.cassandra.ecchronos.core.repair.state.RepairStateImpl.generateSnapshotForVnode(RepairStateImpl.java:134)
        at com.ericsson.bss.cassandra.ecchronos.core.repair.state.RepairStateImpl.generateNewRepairState(RepairStateImpl.java:121)
        at com.ericsson.bss.cassandra.ecchronos.core.repair.state.RepairStateImpl.update(RepairStateImpl.java:79)
        at com.ericsson.bss.cassandra.ecchronos.core.repair.TableRepairJob.postExecute(TableRepairJob.java:229)
        at com.ericsson.bss.cassandra.ecchronos.core.scheduling.ScheduleManagerImpl$JobRunTask.tryRunTask(ScheduleManagerImpl.java:196)
        at com.ericsson.bss.cassandra.ecchronos.core.scheduling.ScheduleManagerImpl$JobRunTask.tryRunTasks(ScheduleManagerImpl.java:184)
        at com.ericsson.bss.cassandra.ecchronos.core.scheduling.ScheduleManagerImpl$JobRunTask.tryRunNext(ScheduleManagerImpl.java:152)
        at com.ericsson.bss.cassandra.ecchronos.core.scheduling.ScheduleManagerImpl$JobRunTask.run(ScheduleManagerImpl.java:140)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
Caused by: javax.naming.ServiceUnavailableException: null
        at com.sun.jndi.rmi.registry.RegistryContext.lookup(RegistryContext.java:136)
        at com.sun.jndi.toolkit.url.GenericURLContext.lookup(GenericURLContext.java:218)
        at javax.naming.InitialContext.lookup(InitialContext.java:417)
        at javax.management.remote.rmi.RMIConnector.findRMIServerJNDI(RMIConnector.java:1955)
        at javax.management.remote.rmi.RMIConnector.findRMIServer(RMIConnector.java:1922)
        at javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:287)
        ... 33 common frames omitted
Caused by: java.rmi.ConnectException: Connection refused to host: localhost; nested exception is: 
        java.net.ConnectException: Connection refused (Connection refused)
        at sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:623)
        at sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:216)
        at sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:202)
        at sun.rmi.server.UnicastRef.newCall(UnicastRef.java:343)
        at sun.rmi.registry.RegistryImpl_Stub.lookup(RegistryImpl_Stub.java:116)
        at com.sun.jndi.rmi.registry.RegistryContext.lookup(RegistryContext.java:132)
        ... 38 common frames omitted
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:607)
        at java.net.Socket.connect(Socket.java:556)
        at java.net.Socket.<init>(Socket.java:452)
        at java.net.Socket.<init>(Socket.java:229)
        at sun.rmi.transport.proxy.RMIDirectSocketFactory.createSocket(RMIDirectSocketFactory.java:40)
        at sun.rmi.transport.proxy.RMIMasterSocketFactory.createSocket(RMIMasterSocketFactory.java:148)
        at sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:617)
        ... 43 common frames omitted
```

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/HostStatesImpl.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/HostStatesImpl.java
@@ -22,6 +22,7 @@ import java.util.concurrent.TimeUnit;
 
 import com.datastax.oss.driver.api.core.metadata.Node;
 import com.ericsson.bss.cassandra.ecchronos.core.utils.DriverNode;
+import com.ericsson.bss.cassandra.ecchronos.core.utils.logging.ThrottlingLogger;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -32,6 +33,7 @@ import org.slf4j.LoggerFactory;
 public final class HostStatesImpl implements HostStates, Closeable
 {
     private static final Logger LOG = LoggerFactory.getLogger(HostStatesImpl.class);
+    private static final ThrottlingLogger THROTTLED_LOGGER = new ThrottlingLogger(LOG, 1, TimeUnit.MINUTES);
 
     private static final long DEFAULT_REFRESH_INTERVAL_IN_MS = TimeUnit.SECONDS.toMillis(10);
 
@@ -129,7 +131,7 @@ public final class HostStatesImpl implements HostStates, Closeable
         }
         catch (IOException e)
         {
-            LOG.warn("Unable to retrieve host states", e);
+            THROTTLED_LOGGER.warn("Unable to retrieve host states", e);
         }
 
         return false;

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/ThrottledLogMessage.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/ThrottledLogMessage.java
@@ -1,0 +1,64 @@
+/*
+ * Copyright 2022 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ericsson.bss.cassandra.ecchronos.core.utils.logging;
+
+import org.slf4j.Logger;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ThrottledLogMessage
+{
+    private final String myMessage;
+    private final long myIntervalNanos;
+    private final AtomicLong myLastLogTime;
+
+    public ThrottledLogMessage(final String message, final long intervalNanos)
+    {
+        myMessage = message;
+        myIntervalNanos = intervalNanos;
+        myLastLogTime = new AtomicLong(Long.MIN_VALUE);
+    }
+
+    private boolean isAllowedToLog(final long timeInNanos)
+    {
+        long lastLogTime = myLastLogTime.get();
+        return timeInNanos >= lastLogTime && myLastLogTime.compareAndSet(lastLogTime, timeInNanos + myIntervalNanos);
+    }
+
+    public final void info(final Logger logger, final long timeInMs, final Object... objects)
+    {
+        if (isAllowedToLog(timeInMs))
+        {
+            logger.info(myMessage, objects);
+        }
+    }
+
+    public final void warn(final Logger logger, final long timeInMs, final Object... objects)
+    {
+        if (isAllowedToLog(timeInMs))
+        {
+            logger.warn(myMessage, objects);
+        }
+    }
+
+    public final void error(final Logger logger, final long timeInMs, final Object... objects)
+    {
+        if (isAllowedToLog(timeInMs))
+        {
+            logger.error(myMessage, objects);
+        }
+    }
+}

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/ThrottlingLogger.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/ThrottlingLogger.java
@@ -1,0 +1,74 @@
+/*
+ * Copyright 2022 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ericsson.bss.cassandra.ecchronos.core.utils.logging;
+
+import org.slf4j.Logger;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Logger that throttles log messages per interval.
+ * A log message uniqueness is based on the string message.
+ *
+ * This logger is thread safe.
+ */
+public class ThrottlingLogger
+{
+    private final Map<String, ThrottledLogMessage> myThrottledLogMessages = new ConcurrentHashMap<>();
+    private final Logger myLogger;
+    private final long myIntervalNanos;
+
+    public ThrottlingLogger(final Logger logger, final long interval, final TimeUnit timeUnit)
+    {
+        myLogger = logger;
+        myIntervalNanos = timeUnit.toNanos(interval);
+    }
+
+    public final void info(final String message, final Object... objects)
+    {
+        ThrottledLogMessage throttledLogMessage = getThrottledLogMessage(message);
+        throttledLogMessage.info(myLogger, System.nanoTime(), objects);
+    }
+
+    public final void warn(final String message, final Object... objects)
+    {
+        ThrottledLogMessage throttledLogMessage = getThrottledLogMessage(message);
+        throttledLogMessage.warn(myLogger, System.nanoTime(), objects);
+    }
+
+    public final void error(final String message, final Object... objects)
+    {
+        ThrottledLogMessage throttledLogMessage = getThrottledLogMessage(message);
+        throttledLogMessage.error(myLogger, System.nanoTime(), objects);
+    }
+
+    private ThrottledLogMessage getThrottledLogMessage(final String message)
+    {
+        ThrottledLogMessage throttledLogMessage = myThrottledLogMessages.get(message);
+        if (throttledLogMessage == null)
+        {
+            throttledLogMessage = new ThrottledLogMessage(message, myIntervalNanos);
+            ThrottledLogMessage addedMessage = myThrottledLogMessages.putIfAbsent(message, throttledLogMessage);
+            if (addedMessage != null)
+            {
+                throttledLogMessage = addedMessage;
+            }
+        }
+        return throttledLogMessage;
+    }
+}

--- a/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/package-info.java
+++ b/core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/package-info.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright 2022 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Utils for logging.
+ */
+package com.ericsson.bss.cassandra.ecchronos.core.utils.logging;

--- a/core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/TestThrottlingLogger.java
+++ b/core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/utils/logging/TestThrottlingLogger.java
@@ -1,0 +1,224 @@
+/*
+ * Copyright 2022 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ericsson.bss.cassandra.ecchronos.core.utils.logging;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.read.ListAppender;
+import org.junit.Test;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestThrottlingLogger
+{
+    private static final String TEST_LOG_MESSAGE = "TEST";
+    private static final IllegalArgumentException TEST_EXCEPTION = new IllegalArgumentException("TEST");
+    private static final long THROTTLE_INTERVAL = 1;
+    private static final TimeUnit THROTTLE_INTERVAL_UNIT = TimeUnit.MINUTES;
+    private static final int DEFAULT_LOG_COUNT = 100;
+
+    @Test
+    public void testInfo()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, THROTTLE_INTERVAL_UNIT);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.info(TEST_LOG_MESSAGE);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.INFO);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testWarn()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, THROTTLE_INTERVAL_UNIT);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.warn(TEST_LOG_MESSAGE);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.WARN);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testError()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, THROTTLE_INTERVAL_UNIT);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.error(TEST_LOG_MESSAGE);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.ERROR);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testInfoWithException()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, THROTTLE_INTERVAL_UNIT);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.info(TEST_LOG_MESSAGE, TEST_EXCEPTION);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.INFO);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testWarnWithException()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, THROTTLE_INTERVAL_UNIT);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.warn(TEST_LOG_MESSAGE, TEST_EXCEPTION);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.WARN);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testErrorWithException()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, THROTTLE_INTERVAL_UNIT);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.error(TEST_LOG_MESSAGE, TEST_EXCEPTION);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.ERROR);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testInfoLowThrottle()
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, TimeUnit.NANOSECONDS);
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            throttlingLogger.info(TEST_LOG_MESSAGE);
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(DEFAULT_LOG_COUNT);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.INFO);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+
+    @Test
+    public void testThreadSafety() throws InterruptedException
+    {
+        Logger unittestLogger = (Logger) LoggerFactory.getLogger(TestThrottlingLogger.class);
+        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+        listAppender.start();
+        unittestLogger.addAppender(listAppender);
+
+        ThrottlingLogger throttlingLogger = new ThrottlingLogger(unittestLogger, THROTTLE_INTERVAL, TimeUnit.MINUTES);
+        CountDownLatch latch = new CountDownLatch(1);
+        Set<Thread> threads = new HashSet<>();
+        for (int i = 0; i < DEFAULT_LOG_COUNT; i++)
+        {
+            threads.add(new Thread(()-> {
+                try
+                {
+                    latch.await();
+                    throttlingLogger.info(TEST_LOG_MESSAGE);
+                }
+                catch (InterruptedException ie)
+                {
+                    //DO NOTHING
+                }
+            }, "TestThread" + i));
+        }
+        // Start all threads at once
+        for (Thread thread : threads)
+        {
+            thread.start();
+        }
+        // Release the latch so all threads can start logging
+        latch.countDown();
+        // Wait for all threads to finish
+        for (Thread thread : threads)
+        {
+            thread.join();
+        }
+
+        List<ILoggingEvent> logs = listAppender.list;
+        assertThat(logs.size()).isEqualTo(1);
+        assertThat(logs.get(0).getLevel()).isEqualTo(Level.INFO);
+        assertThat(logs.get(0).getMessage()).isEqualTo(TEST_LOG_MESSAGE);
+    }
+}

